### ISSUE TITLE
`__CPROVER_havoc_object(p)` havocs wrong object

### ISSUE THREAD

**Issue by remi-delmas-3000** [diffblue/cbmc#7779]
In this program, the havoc statement `__CPROVER_havoc_object(p)` havocs a different object than the one pointed to by `p`.

`p` results from a nondet selection between 4 objects.

Assertions are added to verify that objects not pointed by `p` keep their previous value.

with `BUG=1` assertions related to `obj_1_1` fail, 
with `BUG=2` assertions related to `obj_1_2` fail,  
with `BUG=3` assertions related to `obj_1_3` fail, 
with `BUG=4` assertions related to `obj_1_4` fail.

They fail depending on which object comes first in the ternary operator cascade in the `AO_1_BUG` macro which is used to initialise the field `obj_1_1.next`.
 
The problem goes away as soon if I replace the havoc by an explicit nondet-initialisation:
```c
{
  list_t nondet;
  *p = nondet;
}
```

also goes away if less objects are used in the program

```c
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct list_s {
  int val;
  struct list_s *next;
} list_t;

bool nondet_bool();
int nondet_int();

#define AO_0 &obj_0
#define AO_1                                                                   \
  (nondet_bool()                                                               \
       ? &obj_1_1                                                              \
       : (nondet_bool() ? &obj_1_2 : (nondet_bool() ? &obj_1_3 : &obj_1_4)))

#define BUG 2
#if BUG == 1
#define AO_1_BUG                                                               \
  (nondet_bool()                                                               \
       ? &obj_1_1                                                              \
       : (nondet_bool() ? &obj_1_2 : (nondet_bool() ? &obj_1_3 : &obj_1_4)))
#elif BUG == 2
#define AO_1_BUG                                                               \
  (nondet_bool()                                                               \
       ? &obj_1_2                                                              \
       : (nondet_bool() ? &obj_1_1 : (nondet_bool() ? &obj_1_3 : &obj_1_4)))
#elif BUG == 3
#define AO_1_BUG                                                               \
  (nondet_bool()                                                               \
       ? &obj_1_3                                                              \
       : (nondet_bool() ? &obj_1_2 : (nondet_bool() ? &obj_1_1 : &obj_1_4)))
#else
#define AO_1_BUG                                                               \
  (nondet_bool()                                                               \
       ? &obj_1_4                                                              \
       : (nondet_bool() ? &obj_1_2 : (nondet_bool() ? &obj_1_3 : &obj_1_1)))
#endif

int main() {
  list_t obj_0;

  list_t obj_1_1;
  list_t obj_1_2;
  list_t obj_1_3;
  list_t obj_1_4;

  list_t *p = NULL;
  list_t *list = AO_0;
  list_t *tail = list;

   // Initialises obj_0.next to NULL
  *list = (list_t){.next = NULL, .val = 10};

  // Initialise all next pointers to NULL or the address of &obj_1_i  for some i
  obj_1_1.next = (nondet_bool() ? NULL : AO_1_BUG);
  obj_1_2.next = (nondet_bool() ? NULL : AO_1);
  obj_1_3.next = (nondet_bool() ? NULL : AO_1);
  obj_1_4.next = (nondet_bool() ? NULL : AO_1);

  if (nondet_bool()) {
   // nondet-select an object
    p = AO_1;

    // take snapshots before havoc
    list_t p_old = *p;
    list_t obj_0_old = obj_0;
    list_t obj_1_1_old = obj_1_1;
    list_t obj_1_2_old = obj_1_2;
    list_t obj_1_3_old = obj_1_3;
    list_t obj_1_4_old = obj_1_4;

   // havoc the object
    __CPROVER_havoc_object(p);

    // check that  that objects not pointed by p are unaffected by the havoc (fails unexpectedly)

    if (p != &obj_0) {
      assert(obj_0.val == obj_0_old.val);
      assert(obj_0.next == obj_0_old.next);
    }

    if (p != &obj_1_1) {
      assert(obj_1_1.val == obj_1_1_old.val);
      assert(obj_1_1.next == obj_1_1_old.next);
    }

    if (p != &obj_1_2) {
      assert(obj_1_2.val == obj_1_2_old.val);
      assert(obj_1_2.next == obj_1_2_old.next);
    }

    if (p != &obj_1_3) {
      assert(obj_1_3.val == obj_1_3_old.val);
      assert(obj_1_3.next == obj_1_3_old.next);
    }
    
    if (p != &obj_1_4) {
      assert(obj_1_4.val == obj_1_4_old.val);
      assert(obj_1_4.next == obj_1_4_old.next);
    }

    // initialises the havoced object
    *p = (list_t){.next = NULL, .val = nondet_int()};

    tail->next = p;
    tail = p;

     // so that values are displayed in the trace
    list_t obj_1_1_witness3 = obj_1_1;
    list_t obj_1_2_witness3 = obj_1_2;
    list_t obj_1_3_witness3 = obj_1_3;
    list_t obj_1_4_witness3 = obj_1_4;

    // check heap invariant (expected to pass but fails due to the havoc)
    assert(obj_1_1.next == &obj_1_1 || obj_1_1.next == &obj_1_2 ||
           obj_1_1.next == &obj_1_3 || obj_1_1.next == &obj_1_4 ||
           obj_1_1.next == NULL);

    assert(obj_1_2.next == &obj_1_1 || obj_1_2.next == &obj_1_2 ||
           obj_1_2.next == &obj_1_3 || obj_1_2.next == &obj_1_4 ||
           obj_1_2.next == NULL);

    assert(obj_1_3.next == &obj_1_1 || obj_1_3.next == &obj_1_2 ||
           obj_1_3.next == &obj_1_3 || obj_1_3.next == &obj_1_4 ||
           obj_1_3.next == NULL);

    assert(obj_1_4.next == &obj_1_1 || obj_1_4.next == &obj_1_2 ||
           obj_1_4.next == &obj_1_3 || obj_1_4.next == &obj_1_4 ||
           obj_1_4.next == NULL);
  }
}
```

CBMC version: 5.85
Operating system: MacOS
Exact command line resulting in the issue: cbmc --nondet-static --pointer-check --bounds-check -DBUG=1 abstract_dynamic_object_havoc_bug.c
What behaviour did you expect: all assertions pass
What happened instead: assertions falsified

**Comment by thomasspriggs** [comment]
Quick update from me. I have been able to reproduce this issue and I have managed to substantially reduce it down for easier debugging.
<details>
<summary>reduced.c</summary>

```C
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct list_s {
  struct list_s *next;
} list_t;

bool nondet_bool();

int main() {
  list_t obj_a = {NULL};
  list_t obj_b = {&obj_a};
  list_t obj_c = {NULL};

  list_t *p = nondet_bool() ? &obj_a : (nondet_bool() ? &obj_b : &obj_c);

  // take snapshot before havoc
  list_t obj_a_old = obj_a;

  // havoc the object
  __CPROVER_havoc_object(p);

  __CPROVER_assume(p == &obj_c);
  assert(obj_a.next == obj_a_old.next);
}
```
</details>

<details>
<summary>Trace for the reduced example.</summary>

```
CBMC version 5.88.1 (cbmc-5.88.1-6-gcdea666b9f) 64-bit x86_64 linux
Parsing ./reduced2.c
Converting
Type-checking reduced2
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Adding nondeterministic initialization of static/global variables
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Runtime Symex: 0.00262509s
size of program expression: 65 steps
simple slicing removed 11 assignments
Generated 1 VCC(s), 1 remaining after simplification
Runtime Postprocess Equation: 1.3285e-05s
Passing problem to propositional reduction
converting SSA
Runtime Convert SSA: 0.00109556s
Running propositional reduction
Post-processing
Runtime Post-process: 5.022e-06s
Solving with MiniSAT 2.2.1 with simplifier
1307 variables, 1367 clauses
SAT checker: instance is SATISFIABLE
Runtime Solver: 0.0011827s
Runtime decision procedure: 0.00233821s
Building error trace

** Results:
./reduced2.c function main
[main.assertion.1] line 25 assertion obj_a.next == obj_a_old.next: FAILURE

Trace for main.assertion.1:

State 12 file ./reduced2.c function main line 12 thread 0
----------------------------------------------------
  obj_a.next=((list_t *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 14 file ./reduced2.c function main line 13 thread 0
----------------------------------------------------
  obj_b.next=&obj_a!0@1 (00000100 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 16 file ./reduced2.c function main line 14 thread 0
----------------------------------------------------
  obj_c.next=((list_t *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 25 file ./reduced2.c function main line 16 thread 0
----------------------------------------------------
  p=&obj_c!0@1 (00000010 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 27 file ./reduced2.c function main line 19 thread 0
----------------------------------------------------
  obj_a_old.next=((list_t *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 28 file ./reduced2.c function main line 22 thread 0
----------------------------------------------------
  obj_c={ .next=((list_t *)NULL) } ({ 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 })

State 29 file ./reduced2.c function main line 22 thread 0
----------------------------------------------------
  obj_c.next=((list_t *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 30 file ./reduced2.c function main line 22 thread 0
----------------------------------------------------
  obj_b.next=&obj_a!0@1 (00000100 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 31 file ./reduced2.c function main line 22 thread 0
----------------------------------------------------
  obj_a={ .next=((list_t *)NULL) + 32l } ({ 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00100000 })

State 32 file ./reduced2.c function main line 22 thread 0
----------------------------------------------------
  obj_a.next=((list_t *)NULL) + 32l (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00100000)

Assumption:
  file ./reduced2.c line 24 function main
  p == &obj_c

Violated property:
  file ./reduced2.c function main line 25 thread 0
  assertion obj_a.next == obj_a_old.next
  obj_a.next == obj_a_old.next



** 1 of 1 failed (2 iterations)
VERIFICATION FAILED
```
</details>

The trace clearly shows the havoc instruction updating objects which the pointer does not point to. However I am not much closer to working out why.

**Comment by TGWDB** [comment]
Fixed by #7831
